### DISCUSSION TITLE
Support AdaptiveCPP

### DISCUSSION THREAD

**Discussion by VaiTon** [ggml-org/llama.cpp#10342] [Ideas]
### Feature Description

[AdaptiveCpp](https://github.com/AdaptiveCpp/AdaptiveCpp) is an open source SYCL compiler able to compile SYCL code to bytecode that gets then JIT-compiled on first run, enabling the use of the same binary in different contexts (OpenMP on CPU, SPIRV, CUDA or HIP). 

### Motivation

The Single Pass Compiler provided by AdaptiveCpp is able to JIT compile to different platforms, using different backends. In the long run it could be one of the best alternatives to CUDA for heterogeneous programming.

### Possible Implementation

ggml supports SYCL, but uses a lot of DPC++ specific constructs. They should be abstracted away / removed to adhere to the SYCL standard.

**Comment by tuaris** [comment]
I was going to start a new discussion, but (thankfully) I found this existing one.  I'd also like to express support for this.  Would also like to add that AdaptiveCpp was previously known as hipSYCL or Open SYCL.

This is mostly for the purpose of getting llama.cpp working on FreeBSD using a SYCL backend since AdaptiveCpp is readily available on that platform.

**Comment by henk717** [comment]
@JohannesGaessler they have began adding cuda code compatibility which I do want to let you know about : https://github.com/AdaptiveCpp/AdaptiveCpp/blob/develop/doc/pcuda.md

I assume as is it may not be sufficient, but wanted to let you know so you can have an idea how feasible it is and  so that if you interested its on your radar. Multi vendor with potentially your cuda optimizations for nvidia sounds appealing but I have no clue how workable this would be.

**Reply by JohannesGaessler** [reply]
I think that trying to write generic code such as Vulkan or AdaptiveCpp is fundamentally the wrong approach for GPU computing. The performance is coupled extremely tightly to hardware details so for optimal performance you will need to write hardware-specific code regardless, the question is just whether it's as an ggml backend or as a backend in some other framework.

As of right now I think the hardware with the best value is second-hand NVIDIA GPUs. If that ever changes (e.g. Intel GPUs with 2x 24 GB for $1000) then I will write code specifically for that hardware instead.

**Reply by henk717** [reply]
From what I understood the code does not have to be generic, the main difference is that the optimizations are moved to the runtime loading which generates the ptx instead of compile time. The difference would be that users of an unoptimized GPU would then still have generic code to fall back to, and all GPU's would fall under the same compiler. Its what I was curious about from your perspective, if the approach for applying the GPU specific optimizations in their doc would work or if there would be blockers there.

I'm not saying switch over right away, but lets say that in the future this backend was implemented as a replacement for SYCL and the aim was a universal binary. Would what you do on the CUDA side map in to this ``if(__acpp_sscp_is_device) {...}`` definition or would this way of defining be problematic?

**Reply by JohannesGaessler** [reply]
It doesn't matter whether you have a C equivalent (CUDA) or an assembly equivalent (PTX). You still have to use a hardware-specific memory layout and be careful to stay within the hardware-specific limits of available registers and SRAM. This gets even worse with non-standard instructions such as tensor cores and `__dp4a` (per byte dot product) which are heavily used in the ggml code.

>I'm not saying switch over right away, but lets say that in the future this backend was implemented as a replacement for SYCL and the aim was a universal binary. Would what you do on the CUDA side map in to this if(__acpp_sscp_is_device) {...} definition or would this way of defining be problematic?

That 100% depends on who wants to invest the time and effort for implementation and maintenance. The SYCL and MUSA backends are only a thing because engineers from Intel and Moore Threads are working on them. Unless Intel decides that AdaptiveCpp is a better choice for their hardware SYCL is the backend that will be available for Intel GPUs (unless someone else decides they want an AdaptiveCpp backend). Quite frankly, I will just go with the choice favored by the hardware manufacturer because that will make it the easiest to get support from them if necessary.

I don't know whether my CUDA code would map well into AdaptiveCpp.

**Reply by ghchris2021** [reply]
> I think that trying to write generic code such as Vulkan or AdaptiveCpp is fundamentally the wrong approach for GPU computing. The performance is coupled extremely tightly to hardware details so for optimal performance you will need to write hardware-specific code regardless, the question is just whether it's as an ggml backend or as a backend in some other framework.

I both agree and disagree.  Yes of course if we have computing devices, platforms, ecosystems, accelerators whose performance differs by uh... hmm... 10**7 maybe more then yes, of course, absent much smarter optimizers or optimized / optimizing middleware hardware abstraction layers then of course surely one will have at the application level if not already done at a lower library / toolchain level machine / platform specific performance optimization.

But that's kind of like a reduce to absurdity argument to say that since optimization and adaptation is ultimately necessary then there's no justification to have abstracted high level task focused implementations and the world should just create optimized solutions for each machine and leave out standardized quasi-ubiquitous generic layers which could be SYCL, Vulkan (as you said), OpenCL, OpenMP, even standard C++.

How many layers of optimization, code generation specialization, architecture dependent generic code to optimized code mapping / targeting exist today already in LLVM/Clang, for instance?

Even a "generic personal computer" has lots of variants from AMD64, dozens of optional options like SSE, SSE2/3, MMX, AVX2, AVX512, AMX, probably a dozen others.  ARM NEON, VFP, etc. options.
Bunches of heterogeneous core capabilities within a single CPU.
Hugely prevalent multi-CPU machines with lots of NUMA complexity relevant to optimization.  etc. etc.

So having to use a mapping / targeting / optimizing middleware / toolchain layer (or several stacked!) is status quo normal now even if all you do is optimize for AMD64.

You could not even write a decent BLAS library without having 200 differently optimized / ported variants, but it's still good to have
a generic library / client facing interface that a developer that wants to do LA can program to and themselves not have to worry about platform / machine optimization.

So why not have excellent portable standard programming framework / layer specifications that are machine / platform / accelerator agnostic?  Hiding implementation details, orthogonality, program to interface not to implementation, single responsibility, etc. etc. 

Look at the work going into stdexec, for instance, for C++:

https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p2300r10.html

https://github.com/NVIDIA/stdexec

It seems like it aligns better with SW engineering practice to create and use layers of abstraction that can decouple the business logic domain from the "how you get to ASM code" domain which is why we've stopped programming in ASM mostly and moved to more machine / platform independent abstraction languages like C, C++.

It's even a little absurd that when doing GPGPU / HPC one should have to worry about machine level details to design / implement algorithmic domain code for things like algebra, NN, all that should matter is the code expresses the intent and the result maps to a useful target efficiently.  So for GPGPU why should you even have to 
program inference code differently than you would a generic CPU?

For the most part there is nothing "special" in GPGPU other than data types supported (int, FP), multi-threading, multi-processing, the possibility of distributed computing, NUMA.  Those are exactly the same generic concepts one has to take into account for a generic potato CPU whether amd64, arm, riscv with the noteworthy difference is since
some machines are mostly intended for parallel / vector compute it'd
be a little silly to buy them and program them for serial or other
irrelevant work loads even if the code would still run fine.

Even some FPGA target applications get programmed in OpenCL or such and that's even farther afield than GPUs with compute capability.

**Reply by ghchris2021** [reply]
> It doesn't matter whether you have a C equivalent (CUDA) or an assembly equivalent (PTX). You still have to use a hardware-specific memory layout and be careful to stay within the hardware-specific limits of available registers and SRAM.

IDK I mean it seems to me it has been a long time now that for performance even writing "generic workstation / server code" one HAD to take into account at least the possibilities of the ways the hardware could offer & one could use threading, vectorization to different degrees.  Beyond that even for a very long time now one could not write successfully efficient performance code even serial code and much moreso parallel / NUMA / distributed if one did not realize that multiple layers of cache exist, multiple cores and memory domains (cache, local RAM, less local RAM) exist, and that VM / shared memory exists (sharing / synchronization / latency / ...).

Also memory fetching aspects that typically relate to the caches and those other things so one has to be aware of bandwidth, latency, grouping sequential access to cache lines / cache layer / memory page locality.  The possibility of doing scattered / gathered reads / writes but smartly using tiling or multi-thead synchrony so the overall RAM / cache system efficiently bursts / fills / gets reused / whatever.

So it's not a GPU / accelerator only thing, we're in the middle of that on typical PCs for years if one does want to achieve nearly full performance.

Also look at these big MoE models that can be over 200 GBy in size.  Except for high end GPU servers one would want such inference to work efficiently on not only single processor systems but possibly highly multi-thread, multi-core, multi-CPU server/workstation systems.  One sort of has mandate to assume that running AIML inference workloads should port between using GPGPU and also generic portable programming layers to efficiently target anything from a HEDT to a Mac / EPYC / whatever server with GPU-realm RAM BW and dozens to hundreds of cores and possibly many CPUs.

Local LLM for consumers was/is most often done on GPUs but for the realm of the most interesting (capability of highly complex inference) open models it's mostly about MoE than dense models and very very few people are going to be even using GPUs with 100s of GBy of VRAM for these when DDR5 at 200-1000 GBy/s works great with generic C/C++ code that follows any framework like OpenCL, SYCL, standard C++, OpenMP, etc. etc.
